File: src/api/proyectos.ts
import { callCommand } from './client'
import type {
  Audit,
  EstadoInfo,
  ListQuery,
  LookupItem,
  Money,
  PagedResult,
  RowVersion,
  Uuid,
} from './types'

/** See `docs/11-contratos-tauri.md` §5.3. */

export type EstadoProyecto = 'Activa' | 'Pausada' | 'Finalizada' | 'Cancelada'

export interface ProyectoFiltro {
  /** Matched against the name, the number, the address and the locality. */
  texto?: string
  clienteId?: Uuid
  estado?: EstadoProyecto
  /** Shorthand for the two states that mean "still going on". */
  soloActivas?: boolean
}

/**
 * No state here: it only ever changes through {@link transitionProyecto}, so the form has a single,
 * guarded way in.
 */
export interface ProyectoInput {
  numero: number
  nombre: string
  direccion: string | null
  localidad: string | null
  clienteId: Uuid
}

export interface ProyectoListItem {
  id: Uuid
  numero: number
  nombre: string
  direccion: string | null
  localidad: string | null
  clienteId: Uuid
  clienteNombre: string
  estado: EstadoProyecto
  trabajosCount: number
  rentabilidad: Money
  puedeEliminarse: boolean
  rowVersion: RowVersion
}

export interface ProyectoDetalle {
  id: Uuid
  numero: number
  nombre: string
  direccion: string | null
  localidad: string | null
  clienteId: Uuid
  clienteNombre: string
  estado: EstadoInfo
  trabajosCount: number
  rentabilidad: Money
  puedeEliminarse: boolean
  audit: Audit
}

export const PROYECTOS_SORTABLE = [
  'numero',
  'nombre',
  'clienteNombre',
  'estado',
  'trabajosCount',
  'rentabilidad',
  'createdAt',
] as const

export function listProyectos(query: ListQuery<ProyectoFiltro>): Promise<PagedResult<ProyectoListItem>> {
  return callCommand('proyectos_list', { query })
}

export function getProyecto(id: Uuid): Promise<ProyectoDetalle> {
  return callCommand('proyectos_get', { id })
}

export function createProyecto(dto: ProyectoInput): Promise<ProyectoDetalle> {
  return callCommand('proyectos_create', { dto })
}

export function updateProyecto(id: Uuid, dto: ProyectoInput, rowVersion: RowVersion): Promise<ProyectoDetalle> {
  return callCommand('proyectos_update', { id, dto, rowVersion })
}

/** `cascada` answers the confirmation: close the open jobs along with the site. */
export function transitionProyecto(
  id: Uuid,
  destino: EstadoProyecto,
  rowVersion: RowVersion,
  cascada = false,
): Promise<ProyectoDetalle> {
  return callCommand('proyectos_transition', { id, destino, rowVersion, cascada })
}

export function deleteProyecto(id: Uuid, rowVersion: RowVersion): Promise<void> {
  return callCommand('proyectos_delete', { id, rowVersion })
}

export function lookupProyectos(
  clienteId?: Uuid,
  texto?: string,
  limite?: number,
): Promise<LookupItem[]> {
  return callCommand('proyectos_lookup', { clienteId, texto, limite })
}

/** The number the create form pre-fills with, so the user never has to guess it. */
export function siguienteNumeroProyecto(): Promise<number> {
  return callCommand('proyectos_siguiente_numero', {})
}

File: src-tauri/src/commands/proyectos.rs
//! Commands of `proyectos`. See `docs/11-contratos-tauri.md` §5.5.

use certaro_application::dtos::common::{ListQuery, LookupItem};
use certaro_application::dtos::proyectos::{ProyectoDetalle, ProyectoFiltroDto, ProyectoInput, ProyectoListItem};
use certaro_application::PagedResult;
use certaro_domain::EstadoProyecto;
use tauri::State;
use uuid::Uuid;

use crate::error::{handle, ApiResult};
use crate::state::AppState;

#[tauri::command]
pub async fn proyectos_list(
    state: State<'_, AppState>,
    query: ListQuery<ProyectoFiltroDto>,
) -> ApiResult<PagedResult<ProyectoListItem>> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.list(query).await,
        Err(e) => Err(e),
    };
    handle("proyectos_list", outcome)
}

#[tauri::command]
pub async fn proyectos_get(state: State<'_, AppState>, id: Uuid) -> ApiResult<ProyectoDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.get(id).await,
        Err(e) => Err(e),
    };
    handle("proyectos_get", outcome)
}

#[tauri::command]
pub async fn proyectos_create(state: State<'_, AppState>, dto: ProyectoInput) -> ApiResult<ProyectoDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.create(dto).await,
        Err(e) => Err(e),
    };
    handle("proyectos_create", outcome)
}

#[tauri::command]
pub async fn proyectos_update(
    state: State<'_, AppState>,
    id: Uuid,
    dto: ProyectoInput,
    row_version: String,
) -> ApiResult<ProyectoDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.update(id, dto, &row_version).await,
        Err(e) => Err(e),
    };
    handle("proyectos_update", outcome)
}

/// Moves the site along its state machine. `cascada` is what the confirmation dialog answers: it
/// closes the open jobs along with the site instead of refusing the change.
#[tauri::command]
pub async fn proyectos_transition(
    state: State<'_, AppState>,
    id: Uuid,
    destino: EstadoProyecto,
    row_version: String,
    cascada: Option<bool>,
) -> ApiResult<ProyectoDetalle> {
    let outcome = match state.services() {
        Ok(services) => {
            services
                .proyectos
                .transition(id, destino, &row_version, cascada.unwrap_or(false))
                .await
        }
        Err(e) => Err(e),
    };
    handle("proyectos_transition", outcome)
}

#[tauri::command]
pub async fn proyectos_delete(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.delete(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("proyectos_delete", outcome)
}

#[tauri::command]
pub async fn proyectos_lookup(
    state: State<'_, AppState>,
    cliente_id: Option<Uuid>,
    texto: Option<String>,
    limite: Option<u64>,
) -> ApiResult<Vec<LookupItem>> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.lookup(cliente_id, texto, limite).await,
        Err(e) => Err(e),
    };
    handle("proyectos_lookup", outcome)
}

/// The number the create form pre-fills with, so the user never has to guess it.
#[tauri::command]
pub async fn proyectos_siguiente_numero(state: State<'_, AppState>) -> ApiResult<i32> {
    let outcome = match state.services() {
        Ok(services) => services.proyectos.siguiente_numero().await,
        Err(e) => Err(e),
    };
    handle("proyectos_siguiente_numero", outcome)
}
